addons/mail/models/mail_link_preview.py

```diff
@@ -0,0 +1,160 @@
+        self.env['bus.bus']._sendmany(notifications)
+        self.unlink()
+
+    @api.model
+    def _is_link_preview_enabled(self):
+        link_preview_throttle = int(self.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99))
+        return link_preview_throttle > 0
+
+    @api.model
+    def _is_domain_throttled(self, url):
+        domain = urlparse(url).netloc
+        date_interval = fields.Datetime.to_string((datetime.now() - relativedelta(seconds=10)))
+        call_counter = self.search_count([
+            ('source_url', 'ilike', '%' + domain + '%'),
+            ('create_date', '>', date_interval),
+        ])
+        link_preview_throttle = int(self.env['ir.config_parameter'].sudo().get_param('mail.link_preview_throttle', 99))
+        return call_counter > link_preview_throttle
+
+    @api.model
+    def _get_link_preview_from_url(self, url):
+        try:
+            response = requests.head(url, timeout=3, headers=self.request_headers, allow_redirects=True)
+        except requests.exceptions.RequestException:
+            return False
+        if response.status_code != requests.codes.ok:
+            return False
+        image_mimetype = [
+            'image/bmp',
+            'image/gif',
+            'image/jpeg',
+            'image/png',
+            'image/tiff',
+            'image/x-icon',
+        ]
+        # Content-Type header can return a charset, but we just need the
+        # mimetype (eg: image/jpeg;charset=ISO-8859-1)
+        content_type = response.headers['Content-Type'].split(';')
+        if content_type[0] in image_mimetype:
+            return {
+                'image_mimetype': content_type[0],
+                'source_url': url,
+            }
+        if 'text/html' in response.headers['Content-Type']:
+            return self._get_link_preview_from_html(url)
+        return False
+
+    def _get_link_preview_from_html(self, url):
+        response = requests.get(url, timeout=3, headers=self.request_headers, allow_redirects=True)
+        parser = etree.HTMLParser(encoding=response.encoding)
+        tree = html.fromstring(response.content, parser=parser)
+        og_title = tree.xpath('//meta[@property="og:title"]/@content')
+        if og_title:
+            og_title = og_title[0]
+        if not og_title:
+            return False
+        og_description = tree.xpath('//meta[@property="og:description"]/@content')
+        if og_description:
+            og_description = og_description[0]
+        og_type = tree.xpath('//meta[@property="og:type"]/@content')
+        if og_type:
+            og_type = og_type[0]
+        og_image = tree.xpath('//meta[@property="og:image"]/@content')
+        if og_image:
+            og_image = og_image[0]
+        og_mimetype = tree.xpath('//meta[@property="og:image:type"]/@content')
+        if og_mimetype:
+            og_mimetype = og_mimetype[0]
```

Comment:
Why bother with that if you're going to re-test everything when creating the result?

Just do something like
```python
    'og_mimetype': og_mimetype[0] if og_mimetype else False
```
or
```python
    'og_mimetype': next(iter(og_mimetype, False))
```

And why fallback to `False`? `False` is the fallback for RPC-accessible methods, for compatibility with the XML-RPC protocol. There's no such thing here, use `None`.